### ISSUE TITLE
UTP causes memory leak

### ISSUE THREAD

**Issue by ucwong** [anacrolix/torrent#392]
util got OOM when i use utp enable

**Comment by anacrolix** [comment]
Do you have cgo enabled, and are using anacrolix/go-libutp? Can you provide any profiles? What kind of run time does it take for you to see the leak?

**Comment by ucwong** [comment]
cfg.DisableUTP=false

**Comment by ucwong** [comment]
I think cgo is enable
![image](https://user-images.githubusercontent.com/22344498/78464982-d8e98f00-7722-11ea-9559-27ba76b0050a.png)

**Comment by anacrolix** [comment]
Can you provide any profiles? What kind of run time does it take for you to see the leak?

**Comment by ucwong** [comment]
56161732 56 @ 0xa32e04 0xa32da1 0xf7db26 0x1032e50 0x1028231 0x103b5f2 0x1046ede 0x10471eb 0xa13551
# 0xa32e03 sync.(*Mutex).Unlock+0xa3           /usr/local/go/src/sync/mutex.go:190
# 0xa32da0 sync.(*RWMutex).Unlock+0x40           /usr/local/go/src/sync/rwmutex.go:136
# 0xf7db25 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/sync.(*RWMutex).Unlock+0x55  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/sync/rwmutex.go:26
# 0x1032e4f github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*lockWithDeferreds).Unlock+0x7f /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/deferrwl.go:22
# 0x1028230 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*Client).unlock+0x30  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/client.go:1382
# 0x103b5f1 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*Torrent).AddPeers+0xa1  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/t.go:221
# 0x1046edd github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*trackerScraper).announce+0x58d /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/tracker_scraper.go:126
# 0x10471ea github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*trackerScraper).Run+0x2aa /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/tracker_scraper.go:137


cycles/second=2500063193
1305031203356354 1740 @ 0x9f41c2 0x1047162 0xa13501
# 0x9f41c1 runtime.selectgo+0x521            /usr/local/go/src/runtime/select.go:485
# 0x1047161 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*trackerScraper).Run+0x271 /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/tracker_scraper.go:159

489861550598947 5226 @ 0xa2f7dd 0x102c867 0xa13501
# 0xa2f7dc sync.(*Cond).Wait+0x9c            /usr/local/go/src/sync/cond.go:56
# 0x102c866 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*connection).writer+0x496 /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/connection.go:635

380512478750212 5146 @ 0xa2f7dd 0x10167ad 0x1028f89 0x1036897 0xa5e88c 0xa5e88c 0x1033c30 0xb3b36a 0xa5dc97 0xab8809 0xab8814 0xf32ab3 0x104b7fe 0x102f277 0x1024499 0x1023c83 0x1020da8 0xa13501
# 0xa2f7dc sync.(*Cond).Wait+0x9c             /usr/local/go/src/sync/cond.go:56
# 0x10167ac github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/go-libutp.(*Conn).Read+0x8c   /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/go-libutp/conn.go:159
# 0x1028f88 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.connStatsReadWriter.Read+0x58  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/conn_stats.go:119
# 0x1036896 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*rateLimitedReader).Read+0xb6  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/ratelimitreader.go:43
# 0xa5e88b io.(*multiReader).Read+0xab            /usr/local/go/src/io/multi.go:26
# 0xa5e88b io.(*multiReader).Read+0xab            /usr/local/go/src/io/multi.go:26
# 0x1033c2f github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.deadlineReader.Read+0x12f   /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/handshake.go:27

**Comment by ucwong** [comment]
```
56161732 56 @ 0xa32e04 0xa32da1 0xf7db26 0x1032e50 0x1028231 0x103b5f2 0x1046ede 0x10471eb 0xa13551
# 0xa32e03 sync.(*Mutex).Unlock+0xa3           /usr/local/go/src/sync/mutex.go:190
# 0xa32da0 sync.(*RWMutex).Unlock+0x40           /usr/local/go/src/sync/rwmutex.go:136
# 0xf7db25 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/sync.(*RWMutex).Unlock+0x55  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/sync/rwmutex.go:26
# 0x1032e4f github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*lockWithDeferreds).Unlock+0x7f /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/deferrwl.go:22
# 0x1028230 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*Client).unlock+0x30  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/client.go:1382
# 0x103b5f1 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*Torrent).AddPeers+0xa1  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/t.go:221
# 0x1046edd github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*trackerScraper).announce+0x58d /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/tracker_scraper.go:126
# 0x10471ea github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*trackerScraper).Run+0x2aa /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/tracker_scraper.go:137


cycles/second=2500063193
1305031203356354 1740 @ 0x9f41c2 0x1047162 0xa13501
# 0x9f41c1 runtime.selectgo+0x521            /usr/local/go/src/runtime/select.go:485
# 0x1047161 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*trackerScraper).Run+0x271 /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/tracker_scraper.go:159

489861550598947 5226 @ 0xa2f7dd 0x102c867 0xa13501
# 0xa2f7dc sync.(*Cond).Wait+0x9c            /usr/local/go/src/sync/cond.go:56
# 0x102c866 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*connection).writer+0x496 /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/connection.go:635

380512478750212 5146 @ 0xa2f7dd 0x10167ad 0x1028f89 0x1036897 0xa5e88c 0xa5e88c 0x1033c30 0xb3b36a 0xa5dc97 0xab8809 0xab8814 0xf32ab3 0x104b7fe 0x102f277 0x1024499 0x1023c83 0x1020da8 0xa13501
# 0xa2f7dc sync.(*Cond).Wait+0x9c             /usr/local/go/src/sync/cond.go:56
# 0x10167ac github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/go-libutp.(*Conn).Read+0x8c   /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/go-libutp/conn.go:159
# 0x1028f88 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.connStatsReadWriter.Read+0x58  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/conn_stats.go:119
# 0x1036896 github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.(*rateLimitedReader).Read+0xb6  /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/ratelimitreader.go:43
# 0xa5e88b io.(*multiReader).Read+0xab            /usr/local/go/src/io/multi.go:26
# 0xa5e88b io.(*multiReader).Read+0xab            /usr/local/go/src/io/multi.go:26
# 0x1033c2f github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent.deadlineReader.Read+0x12f   /root/CortexTheseus/build/_workspace/src/github.com/CortexFoundation/CortexTheseus/vendor/github.com/anacrolix/torrent/handshake.go:27
```

**Comment by ucwong** [comment]
Here is some informations I saved before with debug/pprof

**Comment by anacrolix** [comment]
@ucwong Thanks. I'm not sure what's going on there, you have a lot of goroutines blocked on unlocking the Client mutex? What versions of go-libutp and torrent are you on? Could you update to torrent v1.15.0?

**Comment by ucwong** [comment]
OK, I think i used the latest version, but i will test again
it will be ok when disable utp

**Comment by ucwong** [comment]
Still OOM at last in utp mode

**Comment by anacrolix** [comment]
Could it be that you're running *lots* of `Torrent`s, with a lot of connections and trackers? This might be causing a massive amount of contention on the `Client` lock?

**Comment by anacrolix** [comment]
Do you have a read or write rate limiter set?

**Comment by ucwong** [comment]
... But it is ok for tcp protocol only, I use default set ,it maybe unlimited

**Comment by ucwong** [comment]
Any suggestions for utp settings different with tcp

**Comment by anacrolix** [comment]
It's possible there are undocumented, upstream limits to how many connections you can have per utp socket. This information would be helpful:

> Could it be that you're running lots of Torrents, with a lot of connections and trackers? This might be causing a massive amount of contention on the Client lock?

**Comment by ucwong** [comment]
Is UploadRateLimiter in client config OR others

**Comment by james-lawrence** [comment]
@ucwong I have a [fork](https://github.com/james-lawrence/torrent) of this library that massively reworks the locking done. might be worth a look to see it if suffers the same issue. if it doesn't then we can back track possible issues from there potentially.

**Comment by ucwong** [comment]
@james-lawrence thank you for comment, I have test the new commit , the memory seems to be ok now, but the number of block and mutex is still increasing without OOM. I will keep running it for a bit longer time. 
@anacrolix thank you for the latest commits

**Comment by ucwong** [comment]
The memory is increasing slower, I think it still will be OOM after long time seeding
![image](https://user-images.githubusercontent.com/22344498/79532819-e0a51e00-80a8-11ea-851d-0749c34d434d.png)

**Comment by ucwong** [comment]
I found some block here, ratelimitreader.go:43
![image](https://user-images.githubusercontent.com/22344498/79532969-54dfc180-80a9-11ea-8135-465aa7f00378.png)
![image](https://user-images.githubusercontent.com/22344498/79532992-5f9a5680-80a9-11ea-9e27-8bb0d7874ae4.png)

**Comment by ucwong** [comment]
mutex
![image](https://user-images.githubusercontent.com/22344498/79533554-b9e7e700-80aa-11ea-87c0-01c46cab9e17.png)

![image](https://user-images.githubusercontent.com/22344498/79533523-a6d51700-80aa-11ea-83bd-c7d1f3deb4ef.png)

**Comment by anacrolix** [comment]
Are you running with the block and mutex profilers enabled? I'm not sure the results mean what you infer them to mean. I don't have enough information to go on here.

**Comment by ucwong** [comment]
It shows where is blocked in the code, but I am not sure it is useful. It will oom when I seeding files for long time with UTP on. It is very good in only TCP mod.

**Comment by anacrolix** [comment]
A heap profile after it has run for a long time before it crashes would be very helpful.

**Comment by ucwong** [comment]
@anacrolix I will make another test

**Comment by anacrolix** [comment]
@ucwong thanks for sharing that. I would suggest not running with the block or mutex profiling enabled, those won't help with the memory leak as far as I'm aware. How many torrents do you think you have running at a given time? There's a backlog against the dial rate limiter, which I've not seen happen under normal circumstances. The heap profiles also aren't showing anything useful, I'm not sure why. Also what is your CPU usage like? I noticed a lot of CPU devoted to PEX, you might try disabling that.

**Comment by anacrolix** [comment]
@ucwong do you have any more feedback?

**Comment by ucwong** [comment]
I am using it without UTP, It is ok till now. I am not sure why OOM by with UTP seeding

**Comment by ucwong** [comment]
@anacrolix 
![image](https://user-images.githubusercontent.com/22344498/89247890-e2d3be00-d640-11ea-9dd6-ea58daa80f8c.png)

It keeps printing this when I use utp

**Comment by anacrolix** [comment]
@ucwong Does the local system support IPv6?

**Comment by kslr** [comment]
```golang
package main

import (
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"github.com/anacrolix/torrent"
	"log"
	"net/http"
)

func main() {
	config := torrent.NewDefaultClientConfig()
	client, err := torrent.NewClient(config)
	if err != nil {
		log.Panic(client)
	}

	var magnets []string
	for i := 1; i <= 3; i++ {
		url := fmt.Sprintf("https://nyaa.si/?s=seeders&o=desc&p=%d", i)
		response, err := http.Get(url)
		if err != nil {
			log.Fatal(err)
		}
		if response.StatusCode != 200 {
			log.Fatalf("status code error: %d %s", response.StatusCode, response.Status)
		}

		doc, err := goquery.NewDocumentFromReader(response.Body)
		if err != nil {
			log.Fatal(err)
		}
		doc.Find("table > tbody > tr").Each(func(_ int, td *goquery.Selection) {
			td.Find("td").Each(func(i int, a *goquery.Selection) {
				if i == 2 {
					magnet, _ := a.Find("a").Last().Attr("href")
					magnets = append(magnets, magnet)
				}
			})
		})
		_ = response.Body.Close()
	}
	fmt.Printf("all length: %d \n", len(magnets))

	for _, uri := range magnets {
		_, err = client.AddMagnet(uri)
		if err != nil {
			log.Fatal(err)
		}
	}

	select {}
}
```
If UTP is enabled, memory is quickly depleted, If disabled, it takes about 10 minutes to run out. (~1000 torrent)

ubuntu 20.04 / 32G Ram / go 1.6 / lib 1.15.1

**Comment by anacrolix** [comment]
> @ucwong Does the local system support IPv6?

Same question for you @kslr

**Comment by kslr** [comment]
> > @ucwong Does the local system support IPv6?
> 
> Same question for you @kslr 

Yes, ipv4 and ipv6

**Comment by anacrolix** [comment]
Could subscribers to this issue try again with a recent version of anacrolix/torrent?

**Comment by anacrolix** [comment]
@ucwong do you think this is fixed?

**Comment by anacrolix** [comment]
I think I'm observing this issue when I have seeding enabled. Go reports ~85MB of memory in use, but the process is using ~1.25GB.

**Comment by anacrolix** [comment]
A large memory leak was fixed in https://github.com/anacrolix/go-libutp/commit/04303408f4a76e51911c1f5018b58e6d5004c4dd. This is included in anacrolix/torrent [v1.39.1](https://github.com/anacrolix/torrent/releases/tag/v1.39.1) and later.

Further fixes may occur from https://github.com/anacrolix/go-libutp/issues/20, but I expect that the above fix handled most issues. It also includes a lot of performance improvements around uTP handling.
